### PR TITLE
Adding IArgumentCompleterFactory for parameterized ArgumentCompleters

### LINKED ISSUE #12708: Support parameterized argument completers
# Add support for ArgumentCompletionAttributes with parameters 

ArgumentCompleters are in the current implementation created by type, using a default constructor.

This makes it very hard to provide parameters to the completed, limiting their usefulness.

I propose that we add "Factory" support to the completion system, so that an attribute derived from `ArgumentCompleterFactoryAttribute` gets called on a method with a signature like
```csharp
IArgumentCompleter Create();
```

This would allow the derived attribute to use attribute parameters to create the actual argument completer.

Deriving from ArgumentCompleterFactoryAttribute makes it possible to create generic completers, like
```csharp
[DirectoryCompleter(ContainingFile="pswh.exe", Depth=2)]

[DateCompleter(WeekDay='Monday', From="LastYear")]

[GitCommits(Branch='release')]
```

An sample usage could look like this:

```csharp
   /// <summary>
    /// Creates new number completions
    /// </summary>
    public class NumberCompletionsAttribute : ArgumentCompleterFactoryAttribute
    {
        private readonly int _from;
        private readonly int _to;
        private readonly int _step;

        public NumberCompletionsAttribute(int from, int to, int step = 1) 
        {
            _from = @from;
            _to = to;
            _step = step;
        }

        public IArgumentCompleter Create() => new NumberCompleter(_from, _to, _step);
    }

 public class NumberCompleter : IArgumentCompleter
    private readonly int _from;
    private readonly int _to;
    private readonly int _step;

    public NumberCompleter(int from, int to, int step)
    {
        _from = @from;
        _to = to;
        _step = step;
    }

    public IEnumerable<CompletionResult> CompleteArgument(string commandName, string parameterName, string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
        {
/// complete using the passed parameters
        }
    }
```

# Proposed technical implementation details (optional)

There are a couble of ways to go about this:

1. New abstract class `ArgumentCompleterFactoryAttribute` with an abstract `Create` method.
2. New interface `IArgumentCompleterFactory` That the attributes implement if they want the factory support.
3. Combine the above: Have `ArgumentCompleterFactoryAttribute` implement `IArgumentCompleterFactory`

Pull request for implementation can be found here: #12605

### PR DESCRIPTION
**Pull request by powercode**
# PR Summary

Adding support for derived, parameterized, ArgumentCompleterAttributes. This makes it possible to create more generic completers, either in libraries or shipping with the product.

This allows for parameterized ArgumentCompleters. 
Deriving from `ArgumentCompleterAttribute` makes it possible to create generic completers, like
```powershell
[DirectoryCompleter(ContainingFile="pswh.exe", Depth=2)]

[DateCompleter(WeekDay='Monday', From="LastYear")]

[GitCommits(Branch='release')]
```

The derived attributes implement the `IArgumentCompleterFactory` interface and use property values to create a specialized completer.

## PR Context

It is difficult (if even possible) to write generic completers. This PR would allow for the following:

```csharp
   /// <summary>
    /// Creates new number completions
    /// </summary>
    public class NumberCompletionsAttribute : ArgumentCompleterAttribute, IArgumentCompleterFactory
    {
        private readonly int _from;
        private readonly int _to;
        private readonly int _step;

        /// <summary>
        /// Creates a new <see cref="NumberCompletionsAttribute"/>.
        /// </summary>
        /// <param name="from"></param>
        /// <param name="to"></param>
        /// <param name="step"></param>
        public NumberCompletionsAttribute(int from, int to, int step = 1) 
        {
            _from = @from;
            _to = to;
            _step = step;
        }

        /// <inheritdoc />
        public IArgumentCompleter Create() => new NumberCompleter(_from, _to, _step);
    }

    /// <summary>
    /// Completes numbers
    /// </summary>
    public class NumberCompleter : IArgumentCompleter
    {
        private readonly int _from;
        private readonly int _to;
        private readonly int _step;

        /// <summary>
        /// Creates a new instance of a NumberCompleter
        /// </summary>
        /// <param name="from"></param>
        /// <param name="to"></param>
        /// <param name="step"></param>
        public NumberCompleter(int from, int to, int step)
        {
            _from = @from;
            _to = to;
            _step = step;
        }

        /// <inheritdoc />
        public IEnumerable<CompletionResult> CompleteArgument(string commandName, string parameterName, string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
        {
            var res = new List<CompletionResult>((_to - _from) / _step);

            for (int i = _from; i <= _to; i += _step)
            {
                string completionText = i.ToString();
                if (completionText.StartsWith(wordToComplete, StringComparison.Ordinal))
                    res.Add(new CompletionResult(completionText));
            }

            return res;
        }
    }
```

or in PowerShell:
```powershell
using namespace System.Management.Automation
using namespace System.Management.Automation.Language
using namespace System.Collections
using namespace System.Collections.Generic


class NumberCompleter : IArgumentCompleter {

    [int] $From
    [int] $To
    [int] $Step
    
    NumberCompleter([int] $from, [int] $to, [int] $step) {
        if ($from -gt $to) {
            throw [ArgumentOutOfRangeException]::new("from")
        }
        $this.From = $from
        $this.To = $to
        $this.Step = $step -lt 1 ? 1 : $step
    }

    [IEnumerable[CompletionResult]] CompleteArgument(
        [string] $CommandName,
        [string] $parameterName,
        [string] $wordToComplete,
        [CommandAst] $commandAst,
        [IDictionary] $fakeBoundParameters) {
        
        $resultList = [List[CompletionResult]]::new()
        $local:to = $this.To
        $local:step = $this.Step
        for ($i = $this.From; $i -lt $to; $i += $step) {
            $resultList.Add([CompletionResult]::new($i.ToString()))
        }
        
        return $resultList
    }
}

class NumberCompletionAttribute : ArgumentCompleterAttribute, IArgumentCompleterFactory {
    [int] $From
    [int] $To
    [int] $Step
    
    NumberCompletionAttribute([int] $from, [int] $to, [int] $step) {
        $this.From = $from
        $this.To = $to
        $this.Step = $step
    }

    [IArgumentCompleter] Create() { return [NumberCompleter]::new($this.From, $this.To, $this.Step) }
}

```

Usage from PowerShell would then be:

```powershell
function Add{
    param(
       [NumberCompletions(0, 100, 5)]
       [int] $X
      ,
       [NumberCompletions(0, 100, 5)]
       [int] $Y
    )
    $X + $Y
}
```

Fix https://github.com/PowerShell/PowerShell/issues/12708

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: PowerShell.sln.DotSettings [hunk 1]
```diff
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (PoshChan) [inline PowerShell.sln.DotSettings:1]: @powercode, your last commit had 32 failures in `PowerShell-CI-static-analysis`
(These are 5 of the failures)

Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/PowerShell-7.0.0-win-x86.msi should work
```powershell
Failed to complete request to "https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/PowerShell-7.0.0-win-x86.msi". Response status code does not indicate success: 403 (Forbidden).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.1.0-preview.2/PowerShell-7.1.0-preview.2-win-x86.msi should work
```powershell
Failed to complete request to "https://github.com/PowerShell/PowerShell/releases/download/v7.1.0-preview.2/PowerShell-7.1.0-preview.2-win-x86.msi". Response status code does not indicate success: 403 (Forbidden).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/powershell-lts_7.0.0-1.ubuntu.18.04_amd64.deb should work
```powershell
Failed to complete request to "https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/powershell-lts_7.0.0-1.ubuntu.18.04_amd64.deb". Response status code does not indicate success: 403 (Forbidden).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/powershell_7.0.0-1.ubuntu.18.04_amd64.deb should work
```powershell
Failed to complete request to "https://github.com/PowerShell/PowerShell/releases/download/v7.0.0/powershell_7.0.0-1.ubuntu.18.04_amd64.deb". Response status code does not indicate success: 403 (Forbidden).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://github.com/PowerShell/PowerShell/releases/download/v7.1.0-preview.2/powershell-preview_7.1.0-preview.2-1.ubuntu.18.04_amd64.deb should work
```powershell
Failed to complete request to "https://github.com/PowerShell/PowerShell/releases/download/v7.1.0-preview.2/powershell-preview_7.1.0-preview.2-1.ubuntu.18.04_amd64.deb". Response status code does not indicate success: 403 (Forbidden).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```

Comment (iSazonov) [inline PowerShell.sln.DotSettings]: Please remove the extra file.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 2]
```diff
@@ -2030,19 +2030,19 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
             {
                 try
                 {
-                    if (argumentCompleterAttribute.Type != null)
+                    var completer = argumentCompleterAttribute.Type != null
+                        ? Activator.CreateInstance(argumentCompleterAttribute.Type) as IArgumentCompleter
+                        : argumentCompleterAttribute is IArgumentCompleterFactory factory ? factory.Create() : null;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2035]: Interesting, can we put all in a Factory?
I mean something like:
```c#
var completer = Factory.Create(argumentCompleterAttribute)
```

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 3]
```diff
@@ -24,34 +24,41 @@ public class ArgumentCompleterAttribute : Attribute
...
-        public ArgumentCompleterAttribute(ScriptBlock scriptBlock)
+        protected ArgumentCompleterAttribute()
         {
-            if (scriptBlock == null)
+            if (GetType().GetInterfaces().All(t => t != typeof(IArgumentCompleterFactory)))
             {
-                throw PSTraceSource.NewArgumentNullException("scriptBlock");
+                throw PSTraceSource.NewInvalidOperationException();
             }
+        }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: I don't see where it is used.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 4]
```diff
@@ -24,34 +24,41 @@ public class ArgumentCompleterAttribute : Attribute
...
 
-            ScriptBlock = scriptBlock;
+        /// <summary>
+        /// This constructor is used primarily via PowerShell scripts.
+        /// </summary>
+        /// <param name="scriptBlock"></param>
+        public ArgumentCompleterAttribute(ScriptBlock scriptBlock)
+        {
+            ScriptBlock = scriptBlock ?? throw PSTraceSource.NewArgumentNullException("scriptBlock");
         }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:62]: It is unrelated change. Please revert.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: It isn't used yet. This is infrastructure intended to be used by others

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs]: Could you please add more info how should it be used? Maybe the doc comment should be better.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 5]
```diff
@@ -2030,19 +2030,19 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
             {
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2030]: @powercode, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Test PSDesiredStateConfiguration.Get-DscResource - Class base Resources.should be able to get class resource - XmlFileContentResource - Good case
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 348
348:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Class base Resources.should be able to get class resource - XmlFileContentResource - Module Name case mismatch
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 348
348:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Class base Resources.should be able to get class resource - xmlfilecontentresource - Resource name case mismatch
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 348
348:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Class base Resources.should be able to get class resource - XmlFileContentResource from XmlContentDsc - Good case
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 323
323:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Class base Resources.should be able to get class resource - XmlFileContentResource from xmlcontentdsc - Module Name case mismatch
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 323
323:             $resource | Should -Not -BeNullOrEmpty
```

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs]: Good point. Adding an example usage in the doc comments

**Comment by powercode** [comment]
The CI errors seem unrelated to my commits

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 2 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Adding method to ArgumentCompleterAttribute

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs [hunk 6]
```diff
@@ -2030,19 +2030,17 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
...
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2030]: @powercode, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Test PSDesiredStateConfiguration.Get-DscResource - Composite Resources.should be able to get groupset - case mismatch in resource name
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 133
133:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Composite Resources.should be able to get GroupSet - Both names have matching case
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 133
133:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Composite Resources.should be able to get GroupSet - case mismatch in module name
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 133
133:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Composite Resources.should be able to get groupset from PSDscResources - case mismatch in resource name
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 156
156:             $resource | Should -Not -BeNullOrEmpty
```
Test PSDesiredStateConfiguration.Get-DscResource - Composite Resources.should be able to get GroupSet from PSDscResources - Both names have matching case
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/PSDesiredStateConfiguration/PSDesiredStateConfiguration.Tests.ps1: line 156
156:             $resource | Should -Not -BeNullOrEmpty
```

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 7]
```diff
@@ -24,22 +24,34 @@ public class ArgumentCompleterAttribute : Attribute
...
+        /// Initializes a new instance of the <see cref="ArgumentCompleterAttribute"/> class.
+        /// This constructor is used by derived attributes implementing <see cref="IArgumentCompleterFactory"/>.
+        /// </summary>
+        protected ArgumentCompleterAttribute()
+        {
+            if (GetType().GetInterfaces().All(t => t != typeof(IArgumentCompleterFactory)))
+            {
+                throw PSTraceSource.NewInvalidOperationException();
+            }
+        }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: I still do not understand where the constructor is involved.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 8]
```diff
@@ -83,6 +104,57 @@ public interface IArgumentCompleter
             IDictionary fakeBoundParameters);
     }
 
+    /// <summary>
+    /// Creates a new argument completer.
+    /// </summary>
+    /// <para>
+    /// If an attributes that derives from <see cref="ArgumentCompleterAttribute"/> implements this interface,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:111]:
```suggestion
    /// If an attribute that derives from <see cref="ArgumentCompleterAttribute"/> implements this interface,
```

**Comment by powercode** [comment]
Yet again an unrelated CI failure

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 7 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: It is protected to ensure that the only way to call it is from a derived class, and then it fails if the derived class does not implement IArgumentCompleterFactory.

The ctor is implicitly called when the derived class is constructed.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs (hunk 6 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:2030]: This is unrelated to my changes.

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 7 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: Thanks! In the case please add a comment and a test for the scenario.

**Comment by iSazonov** [comment]
Add PowerShell Committee to approve new public API.

**Comment by iSazonov** [comment]
Do you think about IArgumentCompleter<T> vs IArgumentCompleterFactory?

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs [hunk 9]
```diff
@@ -24,22 +24,34 @@ public class ArgumentCompleterAttribute : Attribute
...
             Type = type;
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ArgumentCompleterAttribute"/> class.
+        /// This constructor is used by derived attributes implementing <see cref="IArgumentCompleterFactory"/>.
+        /// </summary>
+        protected ArgumentCompleterAttribute()
+        {
+            if (GetType().GetInterfaces().All(t => t != typeof(IArgumentCompleterFactory)))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:49]:
```suggestion
            if (!typeof(IArgumentCompleterFactory).IsAssignableFrom(this))
```

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 7 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:53]: NumberCompletionAttribute ctor calls it implicitly

File: src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs (hunk 9 above)
Comment (powercode) [reply src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs:49]: Good idea

**Comment by joeyaiello** [comment]
@PowerShell/powershell-committee discussed this and we all think it's super interesting, and, at least in our initial look, a well thought out feature.

However, given it's scope, it'd be really nice to see an RFC (or even an issue) where the problem statement and any alternative solutions are discussed and other contributors have a chance to give feedback on the proposal. 

@JamesWTruher also suggested that it might be useful to see a real-world example of how this could be used to refactor PowerShell built-in completers, or to make them more intelligent. E.g. `Import-Module -Path <tab>` might only show folders, *.psd1 files, and *.psm1 files, and that could be reused in other places where it's relevant.

**Comment by iSazonov** [comment]
Convert to Draft until RFC or issue discussion is finished.

**Comment by daxian-dbw** [comment]
I like the idea, very interesting, allowing people to have reusable argument completer attributes. About the implementation, why not have an abstract attribute class that defines an abstract method which returns an `IArgumentCompleter` instance? For example:

```c#
[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
public abstract class ReusableArgumentCompletionAttribute : Attribute
{
    protected ReusableArgumentCompletionAttribute() {}
    public abstract IArgumentCompleter NewArgumentCompleter(/* EngineIntrinsics engineIntrinsics ?? */);
}
```

This seems more succinct than `: ArgumentCompleterAttribute, IArgumentCompleterFactory`

**Comment by SteveL-MSFT** [comment]
The @PowerShell/powershell-committee discussed this.  In general, we all agree this capability would be useful so thanks for that!  However, we're trying to update our RFC process to make it lighter-weight and would like to try it out with this feature if you're ok with this.  Basically, we'd like to have an issue opened that describes the problem to be solved and/or value to the user and also some example code showing its use and expected results.  Based on the amount and quality of discussion, it will help inform if an RFC is needed.

**Comment by vexx32** [comment]
@SteveL-MSFT in case y'all missed it. 🙂 
https://github.com/PowerShell/PowerShell/issues/12708

**Comment by SteveL-MSFT** [comment]
@vexx32 yes, we missed that.  Thanks!

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee had reviewed the associated [issue](https://github.com/PowerShell/PowerShell/issues/12708#issuecomment-645715122) and supports the current design (which is not a statement on the PR review).

**Comment by powercode** [comment]
@PoshChan Please rebuild all

**Comment by PoshChan** [comment]
@powercode, you are not authorized to request a rebuild

**Comment by powercode** [comment]
🤷‍♂️

**Comment by adityapatwardhan** [comment]
Closing and re-opening as the CI structure has changed

File: test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1 [hunk 10]
```diff
@@ -175,6 +175,77 @@ function TestFunction
...
+}
+
+function FactoryCompletionAdd {
+    param(
+        [NumberCompletion(0, 50, Step = 5)]
+        [int] $Number
+    )
+}
+
+Describe "Factory based extensible completion" -Tags "CI","Factory" {
```
Comment (adityapatwardhan) [inline test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1:239]:
```suggestion
Describe "Factory based extensible completion" -Tags "CI" {
```

Please remove non standard tags.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Comment by adityapatwardhan** [comment]
The CI failure was recently fixed and would be fine on a re-run after the next commit.

**Comment by powercode** [comment]
I think the Codacy error is a false positive.

**Comment by powercode** [comment]
@PowerShellTeam Any eyes on this?

**Comment by powercode** [comment]
@adityapatwardhan It seems that you have a requested change here that I have addressed, but somehow the system does not recognize it. Can you have a check to see if there is something that you still want to be addressed?

**Comment by powercode** [comment]
@PowerShellTeam Can you have a look at why this is in limbo? It's been months without a comment.

**Comment by iSazonov** [comment]
> @PowerShellTeam Can you have a look at why this is in limbo? It's been months without a comment.

175 open PRs - it speaks for itself - MSFT team is working on other projects.

File: test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1 (hunk 10 above)
Comment (powercode) [reply test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1]: Fixed

**Comment by powercode** [comment]
@SteveL-MSFT This PR is in a `Change Requested` state, but the requested change is on a commit that is no longer available due to a force push after a rebase, and the requested change is made. How can I get this into a green state?

**Comment by powercode** [comment]
@adityapatwardhan It's been 8 months since you last touched this. Can you act on it or maybe reassign it if you don't have the bandwidth to process it?

**Comment by powercode** [comment]
@iSazonov I'm suspecting that this is in limbo as there is 1 change requested, but it is requested on a commit that isn't available anymore, and hence cannot be fixed. The requested change is already in a different, force-pushed commit.

What do you think is the right way to move this forward? This is actually a very nice feature to have, that opens up new possibilities, both for the team and for regular users, and it's a shame that it's just stuck here.

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
Sorry for the delay. I have reviewed this and looks good. Since it has been a while since we ran the CI tests for this PR. I am restarting all the tests. I will merge as soon as the tests pass.

**Comment by adityapatwardhan** [comment]
Closing and re-opening to re-trigger all CI tests

**Comment by adityapatwardhan** [comment]
@PoshChan please remind me in 1 hour

**Comment by PoshChan** [comment]
@adityapatwardhan, this is the reminder you requested 1 hour ago

**Comment by adityapatwardhan** [comment]
@powercode Thank you for your contribution and patience!

**Comment by iSazonov** [comment]
@powercode Please open new documentation issue.

/cc @sdwheeler

**Comment by sdwheeler** [comment]
@powercode @iSazonov The scripting example can be added to the docs. The C# usage needs to be added in ///-comments in the source code. Please open the appropriate `documentation needed` issue in this repo.
